up-streamer/src/route.rs

```diff
@@ -0,0 +1,110 @@
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use async_std::sync::{Arc, Mutex};
+use log::*;
+use up_rust::{UAuthority, UTransport};
+
+const ROUTE_TAG: &str = "Route:";
+const ROUTEFN_NEW_TAG: &str = "new():";
+
+///
+/// [`Route`] is defined as a combination of [`UAuthority`][up_rust::UAuthority] and
+/// [`Arc<Mutex<Box<dyn UTransport>>>`][up_rust::UTransport] as routes are at the [`UAuthority`][up_rust::UAuthority] level.
+///
+/// # Examples
+///
+/// ```
+/// use std::sync::Arc;
+/// use async_std::sync::Mutex;
+/// use up_rust::{Number, UAuthority, UTransport};
+/// use up_streamer::Route;
+///
+/// # pub mod up_client_foo {
+/// #     use std::sync::Arc;
+/// #     use up_rust::{UMessage, UTransport, UStatus, UUIDBuilder, UUri, UListener};
+/// #     use async_trait::async_trait;
+/// #     pub struct UPClientFoo;
+/// #
+/// #     #[async_trait]
+/// #     impl UTransport for UPClientFoo {
+/// #         async fn send(&self, _message: UMessage) -> Result<(), UStatus> {
+/// #             todo!()
+/// #         }
+/// #
+/// #         async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+/// #             todo!()
+/// #         }
+/// #
+/// #         async fn register_listener(
+/// #             &self,
+/// #             topic: UUri,
+/// #             _listener: Arc<dyn UListener>,
+/// #         ) -> Result<(), UStatus> {
+/// #             println!("UPClientFoo: registering topic: {:?}", topic);
+/// #             Ok(())
+/// #         }
+/// #
+/// #         async fn unregister_listener(&self, topic: UUri, _listener: Arc<dyn UListener>) -> Result<(), UStatus> {
+/// #             println!(
+/// #                 "UPClientFoo: unregistering topic: {topic:?}"
+/// #             );
+/// #             Ok(())
+/// #         }
+/// #     }
+/// #
+/// #     impl UPClientFoo {
+/// #         pub fn new() -> Self {
+/// #             Self {}
+/// #         }
+/// #     }
+/// # }
+///
+/// let local_transport: Arc<Mutex<Box<dyn UTransport>>> = Arc::new(Mutex::new(Box::new(up_client_foo::UPClientFoo::new())));
+///
+/// let authority_foo = UAuthority {
+///     name: Some("foo_name".to_string()).into(),
+///     number: Some(Number::Ip(vec![192, 168, 1, 100])),
+///     ..Default::default()
+/// };
+///
+/// let local_route = Route::new("local_route", authority_foo, local_transport);
+/// ```
+#[derive(Clone)]
+pub struct Route {
+    pub(crate) name: String,
+    pub(crate) authority: UAuthority,
+    pub(crate) transport: Arc<Mutex<Box<dyn UTransport>>>,
```

Comment:
now that UTransport is `Send + Sync`, do you still need the `Mutex`?

Reply:
I think so yeah. `UTransport` being `Send + Sync` is the "entrypoint" if you will for being able to wrap the trait object in an `Arc<Mutex<>>`.

Reply:
My point is that we do not need the `Mutex` at all anymore.
```rust
pub(crate) transport: Arc<dyn UTransport>,
```
should be enough FMPOV. As long as you make sure that each thread gets its own clone of the Arc, they should be fine/safe to concurrently invoke the transport's functions. That was the overall idea of making it `Send + Sync`, wasn't it?

Reply:
I think we do need the `Mutex` because we're dealing with handling shared mutable state across threads. If `Send + Sync` were not implemented for `UTransport`, then we would be unable to wrap in a `Mutex` to share across threads.

Take these examples:
* [example using `Arc<>`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ca220263f00aca198e5595e38955421f)
* [example using `Arc<Mutex<>>`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=58bcb50769b98d6897649e49ffd7c911)

Given that, here's [another example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59961f2528ef2dd64a99ff8403c354ef) showing that we must have put the trait bounds of `Send + Sync` on `UTransport` in order to then put the trait object within an `Arc<Mutex<>>`.

Reply:
I believe the `Mutex` is not needed as long as `UTransport` is `Send + Sync`.
All [`UTransport`](https://github.com/eclipse-uprotocol/up-rust/blob/3ef940069bacfa744f5055ebac5f358acdc649f0/src/utransport.rs#L132) trait's method take `&self`, this delegates all mutability aspects to the implementer which may use internally one or more `Mutex` or `Atomic`.
So, given [`UTransport`] definition, I believe the `Mutex` here is superfluous since the implementer already needs to take care of mutability, regardless it's in side an `Arc`, `Box`, or any other type of container.

Reply:
Appreciate the feedback Luca & Kai. I rearranged the example I posted to better reflect the trait (taking &self) and handle the internal state using an Arc<Mutex<>> to control access.

It now works with wrapping the trait object just in an Arc<>. Cool!

The updated example:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b2a3b37d94f4f26a95d0be628ad862c

Reply:
I think a more appropriate example would be: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f84e1514dc20cdfcd5eb80b057709c40

I've removed the inner `Arc` in `Counter` and provided an example also for an `AtomicCounter`.

Reply:
@PLeVasseur are you going to change this to `Arc<dyn UTransport>`?

Reply:
Yup, will do so. Been pulled onto another task which is limiting my time a bit. Will try to do so this week.

Feel free to review and leave more feedback. I will address as I am able.

Reply:
FYI @sophokles73, I went ahead and did this.